### PR TITLE
Intacct (custom) - SDK adapter

### PR DESCRIPTION
**Pull request by sharat-developer**
# Features
## Authentication
- custom_auth
_______
## Actions
 - [x] **Add purchase transaction items**
 - [x] Create attachment folder
 - [x] Create attachments
 - [x] Create employee
 - [x] Get api session
 - [x] Get attachment
 - [x] Get attachment folder
 - [x] Get employee
 - [x] Update attachment
 - [x] Update attachment folder
 - [x] Update employee
 - [x] **Update purchase transaction header**
 - [x] **Update purchase transaction items**
_______
 ## More Details
**CS-Contact:**  @vivek88
**Issue link:**  https://github.com/workato/issues/issues/5630

### CODE REVIEW

**Comment by brendonyang** [comment]
Looked through the code and made some minor edits. Edited version can be found here. https://github.com/workato/connector_sdk/pull/119

**Comment by sharat-developer** [comment]
@brendonyang, can you check once?

**Comment by brendonyang** [comment]
Actions are all working fine. Checked for consistency.

File: custom_auth/intacct_connector.rb [hunk 1]
```diff
@@ -0,0 +1,1814 @@
...
+      { name: "login_password", optional: false, control_type: "password" },
+      { name: "sender_id", optional: false },
+      { name: "sender_password", optional: false, control_type: "password" },
+      { name: "company_id", optional: false }
+    ],
+
+    authorization: {
+      type: "custom_auth",
+
+      refresh_on: [/Expired Token/],
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: why is this here? There is no refresh block.

File: custom_auth/intacct_connector.rb [hunk 2]
```diff
@@ -0,0 +1,1814 @@
...
+
+      acquire: ->(_connection) {},
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: why is this needed?

File: custom_auth/intacct_connector.rb [hunk 3]
```diff
@@ -0,0 +1,1814 @@
...
+          current["operation"] = {
+            "authentication" => {
+              "login" => {
+                "userid" => connection["login_username"],
+                "companyid" => connection["company_id"],
+                "password" => connection["login_password"]
+              }
+            }
+          }
+          content = current.delete("content")
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: Why are these 2 lines needed?

File: custom_auth/intacct_connector.rb [hunk 4]
```diff
@@ -0,0 +1,1777 @@
...
+  methods: {
+    parse_nested_elements: lambda { |input|
+      input.map do |key, value|
+        if key == "content!"
+          value
+        else
+          value = value&.first if value.is_a?(Array)
+          { key => call("parse_nested_elements", value) } if value != {}
+        end
+      end&.compact.inject(:merge)
```
[the remark this answers is not in the document]
Comment (eeshansim) [reply custom_auth/intacct_connector.rb]: @sharat-developer

File: custom_auth/intacct_connector.rb [hunk 5]
```diff
@@ -0,0 +1,1814 @@
...
+          { name: "createdby", label: "Created by" },
+          { name: "lastmodified", label: "Last modified" },
+          { name: "lastmodifiedby", label: "Last modified by" }
+        ]
+      }
+    }
+  },
+
+  methods: {
+    parse_nested_elements: lambda { |input|
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: logic here does not handle nested objects, only nested arrays.

File: custom_auth/intacct_connector.rb [hunk 6]
```diff
@@ -0,0 +1,1814 @@
...
+                              dig("response", 0, "operation", 0,
+                                  "result", 0) || {}
+        call("parse_nested_elements", attachment_response)
+      },
+
+      input_fields: lambda { |object_definitions|
+        object_definitions["supdoc_create"]
+      },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 7]
```diff
@@ -0,0 +1,1814 @@
...
+      },
+
+      input_fields: lambda { |object_definitions|
+        object_definitions["supdoc_get"].
+          ignored("creationdate", "createdby", "lastmodified",
+                  "lastmodifiedby", "folder", "description").
+          required("supdocid")
+      },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 8]
```diff
@@ -0,0 +1,1814 @@
...
+      },
+
+     input_fields: lambda { |object_definitions|
+       object_definitions["supdocfolder"].
+         ignored("creationdate", "createdby", "lastmodified",
+                 "lastmodifiedby", "name", "description", "parentfolder").
+         required("supdocfoldername")
+     },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 9]
```diff
@@ -0,0 +1,1814 @@
...
+      },
+
+      input_fields: lambda { |object_definitions|
+        object_definitions["supdocfolder"].
+          ignored("creationdate", "createdby", "lastmodified",
+                  "lastmodifiedby", "name", "description", "parentfolder").
+          required("supdocfoldername")
+      },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 10]
```diff
@@ -0,0 +1,1814 @@
...
+    },
+
+    # API Session related actions
+    get_api_session: {
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: why do we need this action?

File: custom_auth/intacct_connector.rb [hunk 11]
```diff
@@ -0,0 +1,1814 @@
...
+    update_purchase_transaction_header: {
+      subtitle: "Update purchase transaction header",
+      description: "Update <span class='provider'>purchase transaction " \
+        "header</span> in <span class='provider'>Intacct</span>",
+      help: "Pay special attention to enter the values for the " \
+        "fields in the same order as listed below, " \
+        "for the action to be successful!",
+
+      execute: lambda { |_connection, input|
+        build_date_object = lambda { |raw_date|
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: make this a reusable method

File: custom_auth/intacct_connector.rb [hunk 12]
```diff
@@ -0,0 +1,1814 @@
...
+          {
+            "year" => raw_date.to_date.strftime("%Y"),
+            "month" => raw_date.to_date.strftime("%m"),
+            "day" => raw_date.to_date.strftime("%d")
+          }
+        }
+
+        input["datecreated"] = (raw_date = input["datecreated"].presence) &&
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: can be refactored

File: custom_auth/intacct_connector.rb [hunk 13]
```diff
@@ -0,0 +1,1814 @@
...
+        po_txn_response = post("/ia/xml/xmlgw.phtml", payload).
+                          dig("response", 0, "operation", 0, "result", 0) || {}
+        call("parse_nested_elements", po_txn_response)
+      },
+
+      input_fields: lambda { |object_definitions|
+        object_definitions["po_txn_header"].required("@key")
+      },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 14]
```diff
@@ -0,0 +1,1814 @@
...
+        po_txn_response = post("/ia/xml/xmlgw.phtml", payload).
+                          dig("response", 0, "operation", 0, "result", 0) || {}
+        call("parse_nested_elements", po_txn_response)
+      },
+
+      input_fields: lambda { |object_definitions|
+        object_definitions["po_txn_transitem"].required("@key")
+      },
+
+      output_fields: ->(object_definitions) { object_definitions["result"] }
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use better object_definition name

File: custom_auth/intacct_connector.rb [hunk 15]
```diff
@@ -0,0 +1,1814 @@
...
+      subtitle: "Create attachments",
+      description: "Create <span class='provider'>attachments</span> in " \
+        "<span class='provider'>Intacct</span>",
+      help: "Pay special attention to enter the values for the " \
+        "fields in the same order as listed below, " \
+        "for the action to be successful!",
+
+      execute: lambda { |_connection, input|
+        payload = {
+          "content" => {
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: maintain the hierarchy and pre-position the other keys here

```ruby 
payload = {
  "control": {},
  "operation": {
    "authentication": {},
    "content": {
      "function" => {
        "@controlid" => "testControlId",
        "create_supdoc" => input["attachment"]
      }
    }
  }
}
```

Then your `apply` block, can be more efficient:

```ruby
payload do |current_payload|
  current_payload&.[]=(
    "control",
    {
      "senderid": connection["sender_id"],
      "password": connection["sender_password"],
      "controlid": "testControlId",
      "uniqueid": false,
      "dtdversion": 3.0
    })
  current_payload&.[]("operation")&.[]=(
    "authentication",
    {
      "login": {
        "userid": connection["login_username"],
        "companyid": connection["company_id"],
        "password": connection["login_password"]
      }
    })
end
```

File: custom_auth/intacct_connector.rb (hunk 10 above)
Comment (sharat-developer) [reply custom_auth/intacct_connector.rb]: This action will be helpful in some cases, where we need to mimic the page URL of record along with session ID.

File: custom_auth/intacct_connector.rb (hunk 2 above)
Comment (sharat-developer) [reply custom_auth/intacct_connector.rb]: just to enable the `detect_on` block, I have added the empty `acquire` block

**Comment by sharat-developer** [comment]
@brendonyang  as we have refactored the code, can you test the adapter once?

**Comment by brendonyang** [comment]
All action works fine.

File: custom_auth/intacct_connector.rb [hunk 16]
```diff
@@ -0,0 +1,2038 @@
...
+          { name: "lastmodifiedby", label: "Last modified by" }
+        ]
+      }
+    }
+  },
+
+  methods: {
+    parse_xml_to_hash: lambda { |xml_obj|
+      xml_obj["xml"].
+        reject { |key, _value| key[/^@/] }.
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: why reject XML attributes?

File: custom_auth/intacct_connector.rb [hunk 17]
```diff
@@ -0,0 +1,2038 @@
...
+        inject({}) do |hash, (key, value)|
+        if value.is_a?(Array)
+          hash.merge(if value.length != 1 ||
+            xml_obj["array_fields"].include?(key)
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: xml_obj["array_fields"] is repeated, use a variable.

**Comment by brendonyang** [comment]
Works fine.

File: custom_auth/intacct_connector.rb [hunk 18]
```diff
@@ -0,0 +1,2033 @@
...
+            })
+        end
+        format_xml("request")
+      end
+    },
+
+    base_uri: ->(_connection) { "https://api.intacct.com" }
+  },
+
+  test: ->(_connection) {
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: use `lambda do` syntax for multi-line block

File: custom_auth/intacct_connector.rb [hunk 19]
```diff
@@ -0,0 +1,2033 @@
...
+                       {
+                         key => call("parse_xml_to_hash",
+                                     "xml" => value[0],
+                                     "array_fields" => array_fields)
+                       }
+                     end)
+        else
+          value
+        end
+      end&.presence
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: why do you need `&.presence`?

File: custom_auth/intacct_connector.rb [hunk 20]
```diff
@@ -0,0 +1,2033 @@
...
+    end,
+
+    build_date_object: lambda do |date_field|
+      if (raw_date = date_field&.to_date)
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: what happens if user provides an invalid date?

File: custom_auth/intacct_connector.rb [hunk 21]
```diff
@@ -0,0 +1,2033 @@
...
+      description: "Update <span class='provider'>purchase transaction " \
+        "header</span> in <span class='provider'>Intacct</span>",
+      help: "Pay special attention to enter the values for the " \
+        "fields in the same order as listed below, " \
+        "for the action to be successful!",
+
+      execute: lambda do |_connection, input|
+        %w[datecreated dateposted datedue exchratedate].
+          each do |date_field|
+            input&.[]=(date_field,
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: handle invalid date data here with `raise()`

File: custom_auth/intacct_connector.rb [hunk 22]
```diff
@@ -0,0 +1,2033 @@
...
+      }
+      class_response = post("/ia/xml/xmlgw.phtml", payload).
+                       dig("response", 0,
+                           "operation", 0,
+                           "result", 0,
+                           "data", 0) || []
+
+      call("parse_xml_to_hash",
+           "xml" => class_response,
+           "array_fields" => ["class"])["class"]&.
```
Comment (eeshansim) [inline custom_auth/intacct_connector.rb]: is `"class"` key guaranteed to be present? Why do you need to use a safe operator for pluck?

**Comment by brendonyang** [comment]
Works fine, just a thing to note.

@sharat-developer will you be addressing https://github.com/workato/connector_sdk/pull/107/files/9be3fcd0c72bbb1928a48006192827fd8ba18275#r196652091 ?

**Comment by sharat-developer** [comment]
Thanks, @brendonyang.
Our platform itself is handling the invalid date and throws `Invalid date` error.

File: custom_auth/intacct_connector.rb (hunk 19 above)
Comment (sharat-developer) [reply custom_auth/intacct_connector.rb]: This is to remove `{}` for `nil` values, which will get added by inject(*{}*)

**Comment by brendonyang** [comment]
Works fine.
